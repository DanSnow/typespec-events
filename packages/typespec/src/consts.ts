export const PACKAGE_NAME = '@typespec-events/typespec';
