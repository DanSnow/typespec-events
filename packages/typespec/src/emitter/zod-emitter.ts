import {
  type ArrayModelType,
  type BooleanLiteral,
  type EmitContext,
  isArrayModelType,
  isNeverType,
  isNullType,
  isUnknownType,
  isVoidType,
  type Model,
  type NumericLiteral,
  type Program,
  type Scalar,
  type StringLiteral,
  type Tuple,
  type Type,
  type Union,
  type UnionVariant,
} from '@typespec/compiler';
import { camelCase, pascalCase } from 'scule';
import type { EmitterOptions } from '../emitter-options.js';
import { StateKeys } from '../lib.js';
import type { LanguageEmitter } from './framework/emitter-framework.js';

// New function to get the schema name for a given TypeSpec type
function getSchemaName(program: Program, type: Type, namingConvention: 'camelCase' | 'PascalCase'): string | undefined {
  if (type.kind === 'Model') {
    // Check if it's an event model
    const eventName = program.stateMap(StateKeys.isEvent).get(type);
    if (eventName) {
      // It's an event model, use its transformed name
      return `${namingConvention === 'PascalCase' ? pascalCase(eventName) : camelCase(eventName)}Schema`;
    }
    // It's a non-event model, use its own name transformed
    return `${namingConvention === 'PascalCase' ? pascalCase(type.name) : camelCase(type.name)}Schema`;
  }
  // Return undefined for types that don't have a dedicated schema name
  return;
}

// Export the getSchemaName function
export { getSchemaName };

function handleScalarType(_program: Program, type_: Type): string {
  const type = type_ as Scalar;
  switch (type.name) {
    case 'string':
      return 'z.string()';
    case 'int8':
    case 'int16':
    case 'int32':
    case 'int64':
    case 'uint8':
    case 'uint16':
    case 'uint32':
    case 'uint64':
    case 'integer':
      return 'z.number().int()';
    case 'float32':
    case 'float64':
    case 'decimal':
    case 'decimal128':
    case 'number':
      return 'z.number()';
    case 'boolean':
      return 'z.boolean()';
    case 'plainDate':
    case 'plainTime':
    case 'utcDateTime':
    case 'offsetDateTime':
    case 'duration':
    case 'bytes':
      return 'z.string()';
    case 'url':
      return 'z.string().url()';
    default:
      return 'z.unknown()';
  }
}

function handleStringType(_program: Program, type_: Type): string {
  const type = type_ as StringLiteral;
  return `z.literal("${type.value}")`;
}

function handleNumberType(_program: Program, type_: Type): string {
  const type = type_ as NumericLiteral;
  return `z.literal(${type.value})`;
}

function handleBooleanType(_program: Program, type_: Type): string {
  const type = type_ as BooleanLiteral;
  return `z.literal(${type.value})`;
}

function handleTupleType(
  program: Program,
  type_: Type,
  namingConvention: 'camelCase' | 'PascalCase',
  generatedSchemas: Map<string, string>,
): string {
  const type = type_ as Tuple;
  const elementZodStrings = type.values.map((v: Type) =>
    typeSpecTypeToZodString(program, v, namingConvention, generatedSchemas),
  );
  return `z.tuple([${elementZodStrings.join(', ')}])`;
}

function handleArrayType(
  program: Program,
  type_: Type,
  namingConvention: 'camelCase' | 'PascalCase',
  generatedSchemas: Map<string, string>,
): string {
  const type = type_ as ArrayModelType;
  const elementType = type.indexer.value;
  const elementZodString = typeSpecTypeToZodString(program, elementType, namingConvention, generatedSchemas);
  return `z.array(${elementZodString})`;
}

function handleModelType(
  program: Program,
  type_: Type,
  namingConvention: 'camelCase' | 'PascalCase',
  generatedSchemas: Map<string, string>,
): string {
  const type = type_ as Model;

  // Check if this model has already been generated as a separate schema
  const schemaName = getSchemaName(program, type, namingConvention);
  if (schemaName && generatedSchemas.has(schemaName)) {
    return schemaName;
  }

  const properties = type.properties;
  const props: string[] = [];
  for (const prop of properties.values()) {
    const isOptional = prop.optional;
    let propZodString = typeSpecTypeToZodString(program, prop.type, namingConvention, generatedSchemas);
    if (isOptional) {
      propZodString += '.optional()';
    }
    props.push(`${prop.name}: ${propZodString}`);
  }
  return `z.object({ ${props.join(', ')} })`;
}

function handleUnionType(
  program: Program,
  type_: Type,
  namingConvention: 'camelCase' | 'PascalCase',
  generatedSchemas: Map<string, string>,
): string {
  const type = type_ as Union;
  const variants = Array.from(type.variants.values());
  const nonNullVariants = variants.filter((v: UnionVariant) => !isNullType(v.type));
  if (nonNullVariants.length === 0) {
    return 'z.null()';
  }
  const variantZodStrings = nonNullVariants.map((v: UnionVariant) =>
    typeSpecTypeToZodString(program, v.type, namingConvention, generatedSchemas),
  );
  let unionString = `z.union([${variantZodStrings.join(', ')}])`;

  const includesNull = variants.some((v: UnionVariant) => isNullType(v.type));
  if (includesNull) {
    unionString += '.nullable()';
  }
  return unionString;
}

const typeHandlers: Map<
  string,
  (
    program: Program,
    type: Type,
    namingConvention: 'camelCase' | 'PascalCase',
    generatedSchemas: Map<string, string>,
  ) => string
> = new Map([
  ['Model', handleModelType],
  ['Union', handleUnionType],
  ['Scalar', handleScalarType],
  ['String', handleStringType],
  ['Number', handleNumberType],
  ['Boolean', handleBooleanType],
  ['Tuple', handleTupleType],
]);

export function typeSpecTypeToZodString(
  program: Program,
  type: Type,
  namingConvention: 'camelCase' | 'PascalCase',
  generatedSchemas: Map<string, string>,
): string {
  if (isNullType(type)) {
    return 'z.null()';
  }
  if (isNeverType(type)) {
    return 'z.never()';
  }
  if (isUnknownType(type)) {
    return 'z.unknown()';
  }
  if (isVoidType(type)) {
    return 'z.void()';
  }

  // Handle arrays before general models
  if (type.kind === 'Model' && isArrayModelType(program, type)) {
    return handleArrayType(program, type, namingConvention, generatedSchemas);
  }

  // Check if this model has already been generated as a separate schema
  if (type.kind === 'Model') {
    const schemaName = getSchemaName(program, type, namingConvention);
    if (schemaName && generatedSchemas.has(schemaName)) {
      return schemaName; // Return the schema name instead of generating inline
    }
  }

  const handler = typeHandlers.get(type.kind);
  if (handler) {
    return handler(program, type, namingConvention, generatedSchemas);
  }

  // Fallback for unhandled types
  return 'z.unknown()';
}

export class ZodEmitter implements LanguageEmitter {
  private program!: Program;
  private namingConvention: 'camelCase' | 'PascalCase' = 'camelCase';

  init(program: Program, context: EmitContext<EmitterOptions>): void {
    this.program = program;
    const emitterOptions = context.options;
    this.namingConvention = emitterOptions.schemaNamingConvention;
  }

  emit(models: Model[], eventModels: Map<Model, string>): { path: string; content: string } {
    const generatedSchemas = new Map<string, string>();
    const eventMapEntries: string[] = [];

    // Generate schemas for all models
    for (const model of models) {
      this.emitModel(model, generatedSchemas);
    }

    // Generate event map entries
    for (const [model, eventName] of eventModels.entries()) {
      const schemaName = getSchemaName(this.program, model, this.namingConvention);
      if (schemaName) {
        eventMapEntries.push(`  ${JSON.stringify(eventName)}: ${schemaName},`);
      }
    }

    // Combine all generated schemas
    const allSchemas = Array.from(generatedSchemas.values());

    // Generate the combined Zod schema and event map file content
    const combinedFileContent = `import { z } from 'zod';\n
${allSchemas.join('\n\n')}
\n\nexport const eventSchemas = {
${eventMapEntries.join('\n')}
} as const;
`;

    return {
      path: 'events.zod.ts',
      content: combinedFileContent,
    };
  }

  private emitModel(model: Model, generatedSchemas: Map<string, string>) {
    const properties = model.properties;
    const props: string[] = [];
    for (const prop of properties.values()) {
      const isOptional = prop.optional;
      let propZodString = typeSpecTypeToZodString(this.program, prop.type, this.namingConvention, generatedSchemas);
      if (isOptional) {
        propZodString += '.optional()';
      }
      props.push(`${prop.name}: ${propZodString}`);
    }
    const schemaString = `z.object({ ${props.join(', ')} })`;

    // Get the schema name and add to generated schemas
    const schemaName = getSchemaName(this.program, model, this.namingConvention);
    if (schemaName) {
      generatedSchemas.set(schemaName, `export const ${schemaName} = ${schemaString};`);
    }
  }
}
