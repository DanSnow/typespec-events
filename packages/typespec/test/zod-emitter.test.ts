import { describe, expect, it } from 'vitest';
import { PACKAGE_NAME } from '../src/consts.js';
import { typeSpecTypeToZodString } from '../src/emitter/zod-emitter.js';
import { assertDefined, compileTypeSpec, emit } from './test-host.js';

it('emit events.zod.ts with camelCase (default)', async () => {
  const results = await emit('@event("cta_clicked") model CtaClicked { title: string}');
  expect(results['events.zod.ts']).toMatchSnapshot();
});

it('emit events.zod.ts with PascalCase', async () => {
  const results = await emit('@event("cta_clicked") model CtaClicked { title: string}', {
    options: {
      [PACKAGE_NAME]: {
        schemaNamingConvention: 'PascalCase',
      },
    },
  });
  expect(results['events.zod.ts']).toMatchSnapshot();
});

describe('typeSpecTypeToZodString', () => {
  it('should handle simple model', async () => {
    const { program, diagnostics } = await compileTypeSpec('model SimpleModel { name: string, age: integer }');
    expect(diagnostics.length).toBe(0);
    const model = program.resolveTypeReference('SimpleModel')[0];
    assertDefined(model);
    expect(typeSpecTypeToZodString(program, model, 'camelCase', new Map())).toBe(
      'z.object({ name: z.string(), age: z.number().int() })',
    );
  });

  it('should handle model with optional properties', async () => {
    const { program, diagnostics } = await compileTypeSpec('model OptionalModel { name?: string, age?: integer }');
    expect(diagnostics.length).toBe(0);
    const model = program.resolveTypeReference('OptionalModel')[0];
    assertDefined(model);
    expect(typeSpecTypeToZodString(program, model, 'camelCase', new Map())).toBe(
      'z.object({ name: z.string().optional(), age: z.number().int().optional() })',
    );
  });

  it('should handle nested models', async () => {
    const { program, diagnostics } = await compileTypeSpec(`
      model Address { street: string, city: string }
      model User { name: string, address: Address }
    `);
    expect(diagnostics.length).toBe(0);
    const userModel = program.resolveTypeReference('User')[0];
    assertDefined(userModel);
    expect(typeSpecTypeToZodString(program, userModel, 'camelCase', new Map())).toBe(
      'z.object({ name: z.string(), address: z.object({ street: z.string(), city: z.string() }) })',
    );
  });

  it('should handle arrays of primitive types', async () => {
    const { program, diagnostics } = await compileTypeSpec('model ArrayModel { names: string[], ages: integer[] }');
    expect(diagnostics.length).toBe(0);
    const model = program.resolveTypeReference('ArrayModel')[0];
    assertDefined(model);
    expect(typeSpecTypeToZodString(program, model, 'camelCase', new Map())).toBe(
      'z.object({ names: z.array(z.string()), ages: z.array(z.number().int()) })',
    );
  });

  it('should handle arrays of models', async () => {
    const { program, diagnostics } = await compileTypeSpec(`
      model Item { id: string }
      model ItemList { items: Item[] }
    `);
    expect(diagnostics.length).toBe(0);
    const model = program.resolveTypeReference('ItemList')[0];
    assertDefined(model);
    expect(typeSpecTypeToZodString(program, model, 'camelCase', new Map())).toBe(
      'z.object({ items: z.array(z.object({ id: z.string() })) })',
    );
  });

  it('should handle unions', async () => {
    const { program, diagnostics } = await compileTypeSpec('model UnionModel { value: string | integer | boolean }');
    expect(diagnostics.length).toBe(0);
    const model = program.resolveTypeReference('UnionModel')[0];
    assertDefined(model);
    expect(typeSpecTypeToZodString(program, model, 'camelCase', new Map())).toBe(
      'z.object({ value: z.union([z.string(), z.number().int(), z.boolean()]) })',
    );
  });

  it('should handle nullable unions', async () => {
    const { program, diagnostics } = await compileTypeSpec(
      'model NullableUnionModel { value: string | null | integer }',
    );
    expect(diagnostics.length).toBe(0);
    const model = program.resolveTypeReference('NullableUnionModel')[0];
    assertDefined(model);
    expect(typeSpecTypeToZodString(program, model, 'camelCase', new Map())).toBe(
      'z.object({ value: z.union([z.string(), z.number().int()]).nullable() })',
    );
  });

  it('should handle tuples', async () => {
    const { program, diagnostics } = await compileTypeSpec('model TupleModel { coords: [string, integer, boolean] }');
    expect(diagnostics.length).toBe(0);
    const model = program.resolveTypeReference('TupleModel')[0];
    assertDefined(model);
    expect(typeSpecTypeToZodString(program, model, 'camelCase', new Map())).toBe(
      'z.object({ coords: z.tuple([z.string(), z.number().int(), z.boolean()]) })',
    );
  });

  it('should handle literal types', async () => {
    const { program, diagnostics } = await compileTypeSpec(
      `model LiteralModel { status: "success" | "failure", code: 200 | 500, enabled: true }`,
    );
    expect(diagnostics.length).toBe(0);
    const model = program.resolveTypeReference('LiteralModel')[0];
    assertDefined(model);
    expect(typeSpecTypeToZodString(program, model, 'camelCase', new Map())).toBe(
      'z.object({ status: z.union([z.literal("success"), z.literal("failure")]), code: z.union([z.literal(200), z.literal(500)]), enabled: z.literal(true) })',
    );
  });

  it('should handle various scalar types', async () => {
    const { program, diagnostics } = await compileTypeSpec(
      'model ScalarModel { s: string, i: integer, f: float32, b: boolean, d: plainDate, u: url }',
    );
    expect(diagnostics.length).toBe(0);
    const model = program.resolveTypeReference('ScalarModel')[0];
    assertDefined(model);
    expect(typeSpecTypeToZodString(program, model, 'camelCase', new Map())).toBe(
      'z.object({ s: z.string(), i: z.number().int(), f: z.number(), b: z.boolean(), d: z.string(), u: z.string().url() })',
    );
  });

  it('emit nested models in correct order', async () => {
    const results = await emit(`
      model City { name: string, zip: int32 }
      model Address { street: string, city: City }
      @event("user_address_updated")
      model UserAddressUpdated { userId: string, address: Address }
    `);
    expect(results['events.zod.ts']).toMatchSnapshot();
  });
});
