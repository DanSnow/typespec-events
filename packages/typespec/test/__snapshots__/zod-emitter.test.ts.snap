// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`emit events.zod.ts with PascalCase 1`] = `
"import { z } from 'zod';

export const CtaClickedSchema = z.object({ title: z.string() });


export const eventSchemas = {
  "cta_clicked": CtaClickedSchema,
} as const;
"
`;

exports[`emit events.zod.ts with camelCase (default) 1`] = `
"import { z } from 'zod';

export const ctaClickedSchema = z.object({ title: z.string() });


export const eventSchemas = {
  "cta_clicked": ctaClickedSchema,
} as const;
"
`;

exports[`typeSpecTypeToZodString > emit nested models in correct order 1`] = `
"import { z } from 'zod';

export const citySchema = z.object({ name: z.string(), zip: z.number().int() });

export const addressSchema = z.object({ street: z.string(), city: citySchema });

export const userAddressUpdatedSchema = z.object({ userId: z.string(), address: addressSchema });


export const eventSchemas = {
  "user_address_updated": userAddressUpdatedSchema,
} as const;
"
`;
